features/192-asset-gear/index.md

```diff
@@ -0,0 +1,495 @@
+
+##### Request Example
+
+```json
+{
+  "content_type": "image/png",
+  "content_md5": "Zum77CZrrrGRDM18nlplig==",
+  "content_length": 26419
+}
+```
+
+#### Response JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "asset_id": { "type": "string" },
+    "url": { "type": "string" },
+    "method": { "type": "string" },
+    "headers": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "name": { "type": "string" },
+          "value": { "type": "string" }
+        }
+      }
+    }
+  },
+  "required": ["asset_id", "url", "method", "headers"]
+}
+```
+
+- `asset_id`: The asset ID. It becomes valid when the asset is uploaded.
+- `url`: The presigned URL to upload the asset.
+- `method`: The method to be used in the upload request.
+- `headers`: The headers to be included in the upload request.
+
+##### Response Example
+
+```json
+{
+  "asset_id": "myimage.png",
+  "url": "https://storage.skygearapps.com/bucket/myappid/myimage.png?...",
+  "method": "PUT",
+  "headers": [
+    { "name": "content-type", "value": "image/png" },
+    { "name": "content-md5", "value": "Zum77CZrrrGRDM18nlplig==" },
+    { "name": "content-length", "value": "26419" }
+  ]
+}
+```
+
+#### Specification
+
+1. Validate `content_type`.
+1. Let `name` be a randomly generated string.
+1. Let `ext` be the file extension derived from `content_type`.
+1. If `ext` is found, append it to `name`.
+1. Let `asset_id` be `/<app-id>/<name>`.
+1. Presign `asset_id` with `content_type`, `content_length`, `content_md5` as headers.
+1. Return the presigned request.
+
+### POST /_asset/sign
+
+#### Description
+
+This endpoint signs the given assets.
+
+This endpoint requires Master Key.
+
+#### Request JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "assets": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "asset_id": { "type": "string" }
+        },
+        "required": ["asset_id"]
+      }
+    }
+  },
+  "required": ["assets"]
+}
+```
+
+- `asset_id`: The asset ID.
+
+##### Request Example
+
+```json
+{
+  "assets": [
+    { "asset_id": "myimage.png" }
+  ]
+}
+```
+
+#### Response JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "assets": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "asset_id": { "type": "string" },
+          "url": { "type": "string" }
+        },
+        "required": ["asset_id", "url"]
+      }
+    }
+  },
+  "required": ["assets"]
+}
+```
+
+- `url`: The signed URL.
+
+##### Response Example
+
+```json
+{
+  "assets": [
+    {
+      "asset_id": "myimage.png",
+      "url": "https://myappname.skygearapps.com/_asset/get/myimage.png?..."
+    }
+  ]
+}
+```
+
+#### Specification
+
+1. Sign each `asset_id` in `assets`.
+1. Return the result in the original order.
+
+### GET /_asset/get/<asset_id>
```

Comment:
I don't think we need to `get` path segment; asset ID is generated randomly and controlled by us, so we can make sure the asset ID would not collide with our API endpoints. Also, it looks weird to have a verb in this URL.

Reply:
As long as `gorilla/mux` supports this, I am fine with removing `/get`

Reply:
```golang
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"

	"github.com/gorilla/mux"
)

func main() {
	router := mux.NewRouter()
	router.HandleFunc("/_asset/sign", func(w http.ResponseWriter, r *http.Request) {
		fmt.Printf("sign\n")
	})
	router.HandleFunc("/_asset/{asset_id}", func(w http.ResponseWriter, r *http.Request) {
		fmt.Printf("asset_id\n")
	})
	w := httptest.NewRecorder()
	r, _ := http.NewRequest("GET", "http://example.com/_asset/sign", nil)
	router.ServeHTTP(w, r)
}

```

The order of registration is important. `/_asset/sign` must be registered first.